driver-core/src/main/com/mongodb/operation/MixedBulkWriteOperation.java

```diff
@@ -211,423 +226,443 @@ public void call(final AsyncConnection connection, final Throwable t) {
+                                } else {
+                                    if (LOGGER.isDebugEnabled()) {
+                                        LOGGER.debug(format("Asynchronously received response for batch %d", batchMetadata.batchNumber));
+                                    }
 
-                @Override
-                BulkWriteResult executeWriteCommandProtocol() {
-                    return connection.deleteCommand(namespace, ordered, writeConcern, deleteRequests, sessionContext);
-                }
+                                    processCommandResult(connection.getDescription().getServerAddress(), bulkWriteBatchCombiner,
+                                            batchMetadata, result);
 
-                @Override
-                WriteRequest.Type getType() {
-                    return DELETE;
+                                    executeBatchesAsync(connection, sessionContext, bulkWriteBatchCombiner, batchMetadata
+                                            .getNextBatchMetadata(), callback);
+                                }
+                            }
+                        });
+            } else {
+                if (bulkWriteBatchCombiner.hasErrors()) {
+                    callback.onResult(null, bulkWriteBatchCombiner.getError());
+                } else {
+                    callback.onResult(bulkWriteBatchCombiner.getResult(), null);
                 }
-            };
+            }
+        } catch (Throwable t) {
+            callback.onResult(null, t);
         }
+    }
 
-        @SuppressWarnings("unchecked")
-        RunExecutor getInsertsRunExecutor(final List<InsertRequest> insertRequests, final Boolean bypassDocumentValidation,
-                                          final Connection connection, final SessionContext sessionContext) {
-            return new RunExecutor(connection.getDescription()) {
+    private void executeLegacyBatchesAsync(final AsyncConnection connection, final List<? extends WriteRequest> writeRequests,
+                                           final int batchNum, final SingleResultCallback<BulkWriteResult> callback) {
+        try {
+            if (!writeRequests.isEmpty()) {
+                WriteRequest writeRequest = writeRequests.get(0);
+                final List<? extends WriteRequest> remaining = writeRequests.subList(1, writeRequests.size());
 
-                @Override
-                void executeWriteProtocol(final int index) {
-                    connection.insert(namespace, ordered, writeConcern, singletonList(insertRequests.get(index)));
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(format("Asynchronously sending batch %d", batchNum));
                 }
 
-                @Override
-                BulkWriteResult executeWriteCommandProtocol() {
-                    return connection.insertCommand(namespace, ordered, writeConcern, bypassDocumentValidation, insertRequests,
```

Comment:
I see that you didn't remove any of the unused code under insertCommand/updateCommand/deleteCommand.  While it would be a backward compatibility break to do so, I also highly doubt anyone is using Connection or AsyncConnection directly and therefore there is not much risk in doing it.

What do you think?

Reply:
I have deleted them in the  [2570Del branch](https://github.com/rozza/mongo-java-driver/tree/j2570Del). 

Its our own internal implementation of the connection that is the gateway, so I think we have the right to delete it.

Reply:
Except that Connection/AsyncConnection are public

Reply:
Thats true, and the API is deprecated, I believe we are entitled to change our implementation.

Reply:
Added some changes to the [2570Del branch](https://github.com/rozza/mongo-java-driver/tree/j2570Del) - that enables the deprecated API.  Its quite ugly but means we won't lose any support for the commands.

Reply:
I think we're entitled only because I am 99.9% sure there are no applications using Connection/AsyncConnection directly.